#
# Submitted-By: Sascha Hauer, 2007-04-20
# Committed-By: Sascha Hauer
#
# Error:
#
# off-by-one error in parameter parsing in stty
#
# State:
#
# fixed in busybox svn since 2007-04-20
#

---
 coreutils/stty.c |   28 +++++++++++++++-------------
 1 file changed, 15 insertions(+), 13 deletions(-)

Index: busybox-1.4.2/coreutils/stty.c
===================================================================
--- busybox-1.4.2.orig/coreutils/stty.c
+++ busybox-1.4.2/coreutils/stty.c
@@ -545,32 +545,33 @@ static const struct control_info *find_c
 
 enum {
 	param_need_arg = 0x80,
-	param_line   = 1 | 0x80,
-	param_rows   = 2 | 0x80,
-	param_cols   = 3 | 0x80,
-	param_size   = 4,
-	param_speed  = 5,
-	param_ispeed = 6 | 0x80,
-	param_ospeed = 7 | 0x80,
+	param_line    = 1 | 0x80,
+	param_rows    = 2 | 0x80,
+	param_cols    = 3 | 0x80,
+	param_columns = 4 | 0x80,
+	param_size    = 5,
+	param_speed   = 6,
+	param_ispeed  = 7 | 0x80,
+	param_ospeed  = 8 | 0x80,
 };
 
 static int find_param(const char * const name)
 {
-	const char * const params[] = {
+	static const char * const params[] = {
 		"line",
 		"rows",
 		"cols",
 		"columns",
-		"size",
-		"speed",
+		"size",  /* 5 */
+		"speed", /* 6 */
 		"ispeed",
 		"ospeed",
 		NULL
 	};
-	int i = index_in_str_array(params, name);
-	if (i < 0)
+	int i = index_in_str_array(params, name) + 1;
+	if (i == 0)
 		return 0;
-	if (!(i == 4 || i == 5))
+	if (!(i == 5 || i == 6))
 		i |= 0x80;
 
 	return i;
@@ -1008,6 +1009,7 @@ end_option:
 #ifdef TIOCGWINSZ
 		case param_rows:
 		case param_cols:
+		case param_columns:
 			xatoul_range_sfx(argnext, 1, INT_MAX, stty_suffixes);
 			break;
 		case param_size:
